tensorflow/compiler/xla/service/gpu/miopen_conv_algorithm_picker.cc

```diff
@@ -77,6 +76,65 @@ StatusOr<se::DeviceMemory<uint8>> ScratchAllocator::AllocateBytes(
   return se::DeviceMemory<uint8>(buffer_addr);
 }
 
+StatusOr<se::dnn::ConvolutionKind> GetDnnConvolutionKind(
+    const HloCustomCallInstruction* conv) {
+  TF_ASSIGN_OR_RETURN(CudnnConvKind kind, GetCudnnConvKind(conv));
+
+  switch (kind) {
+    case CudnnConvKind::kBackwardFilter:
+      return se::dnn::ConvolutionKind::BACKWARD_FILTER;
+    case CudnnConvKind::kBackwardInput:
+      return se::dnn::ConvolutionKind::BACKWARD_DATA;
+    case CudnnConvKind::kForward:
+      return se::dnn::ConvolutionKind::FORWARD;
+    default:
+      break;
+  }
+  return InternalError("Unsupported convolution type : %s", conv->ToString());
+}
+
+StatusOr<se::dnn::DataType> GetDnnDataType(
```

Comment:
Record it here because once I rebase to current `develop-upstream` I will lose this information. This function will cause the debug build fail to link because of the use of `se::dnn::ToDataType<>::value`. I think this is a bug in hcc, the way around it is to add definition as below (before the function). 

>  template <>
    constexpr se::dnn::DataType se::dnn::ToDataType<float>::value;
    template <>
    constexpr se::dnn::DataType se::dnn::ToDataType<double>::value;
    template <>
    constexpr se::dnn::DataType se::dnn::ToDataType<Eigen::half>::value;


To give some details about the original complaints from debug build. It errors out on `ToDataType::value` declared but not defined. This should be a false positive because the `value` field is declared with `static constexpr` which should allow the static field to be used immediately after declaration.

Note that adding definition to where `ToDataType` (in stream executor) won't work. Because stream executor is pulled in multiple times, this results in another linker issue complaining multiple definition to it.